### PR TITLE
Initial support for secondary scheduler extra resources

### PR DESCRIPTION
**Pull request by ffromani**
Add support in the reconciler code to fetch secondary-scheduler extra resources using the `/bin/get-manifests` entrypoint and an auxiliary deployment.

For details, see: https://github.com/openshift/secondary-scheduler-operator/issues/9#issuecomment-992675873

### CODE REVIEW

**Comment by fromanirh** [comment]
the code works as intended (tested on a local cluster). I'd love to add some tests, thinking about it.

**Review by ingvagabund — COMMENTED** [review]
Thank you for bringing this idea. Could you alongside the implementation provide a sample of the logs that are parsed? Is it a single log line? What is the log line(s) format? Plain text/json/other format?

**Comment by fromanirh** [comment]
> Thank you for bringing this idea. Could you alongside the implementation provide a sample of the logs that are parsed? Is it a single log line? What is the log line(s) format? Plain text/json/other format?

Sure thing. The output to be consumed should be kube manifests of the supported objects (ConfigMap, ClusterRoleBinding, ClusterRole) in JSON format. I'll make sure this requirement is more evident. An [example](https://github.com/openshift-kni/scheduler-plugins) I was working on (and which drove the need for this feature)
```json
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: topology-aware-scheduler
rules:
- apiGroups: ["topology.node.k8s.io"]
  resources: ["noderesourcetopologies"]
  verbs: ["get", "list", "watch"]
---
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: topology-aware-scheduler
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: topology-aware-scheduler
subjects:
- kind: ServiceAccount
  name: placeholder
  namespace: placeholder
```

The `get-manifests` entry point can be something as simple as:
```bash
#!/bin/sh
cat << EOF
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: topology-aware-scheduler
rules:
- apiGroups: ["topology.node.k8s.io"]
  resources: ["noderesourcetopologies"]
  verbs: ["get", "list", "watch"]
---
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: topology-aware-scheduler
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: topology-aware-scheduler
subjects:
- kind: ServiceAccount
  name: placeholder
  namespace: placeholder
EOF
/bin/sleep inf
```

File: pkg/operator/extra_resources.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+			if required.Spec.Template.Spec.Containers[i].Image == pat {
+				required.Spec.Template.Spec.Containers[i].Image = img
+				break
+			}
+		}
+	}
+
+	// TODO
+	required.Spec.Replicas = newInt32(1)
+	// FIXME: this method will disappear in 4.6 so we need to fix this ASAP
```
[the remark this answers is not in the document]
Comment (fromanirh) [reply pkg/operator/extra_resources.go:63]: Will have a look!

File: pkg/apis/secondaryscheduler/v1/types.go [hunk 2]
```diff
@@ -33,6 +33,9 @@ type SecondarySchedulerSpec struct {
 
 	// SchedulerImage sets the container image url to be pulled for the custom scheduler that's deployed
 	SchedulerImage string `json:"imageSpec"`
+
+	// ExtraResources informs the operator to pull out extra resources from the SchedulerImage
+	ExtraResources bool `json:"extraResources"`
```
Comment (fromanirh) [inline pkg/apis/secondaryscheduler/v1/types.go:38]: should probably be `omitempty`

File: bindata/assets/secondary-scheduler/deployment-extra.yaml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        app: "secondary-scheduler-extra-resources"
+    spec:
+      restartPolicy: "Always"
+      containers:
+        - name: "secondary-scheduler-extra-resources"
+          image: ${IMAGE}
+          resources:
+            limits:
+              cpu: "100m"
+              memory: "500Mi"
```
Comment (fromanirh) [inline bindata/assets/secondary-scheduler/deployment-extra.yaml:25]: we can probably lower this like a lot considering the purpose of the endpoint. Also not sure about the QoS class.

**Pull request closed without merging**